firebase-common/src/main/java/com/google/firebase/concurrent/LimitedConcurrencyExecutor.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.concurrent;
+
+import androidx.annotation.GuardedBy;
+import com.google.firebase.components.Preconditions;
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingQueue;
+
+class LimitedConcurrencyExecutor implements Executor {
+
+  private final Executor delegate;
+
+  @GuardedBy("queue")
+  private int activeCount;
```

Comment:
would it make sense to use `Semaphore` here instead?

Reply:
Good call, will switch to that